### ISSUE TITLE
Pmono(Artic) broken on current 3.9

### ISSUE THREAD

**Issue by patrickdupuis** [supercollider/supercollider#3231]
Pmono and PmonoArtic are broken in current 3.9. A recent PR seems to have broken things 6c56b0167312cc759c9ab62cd4b242bf7dbd2143 @adcxyz could you please have a look? 

### Error Message
```
ERROR: Message 'defaultGroup' not understood.
RECEIVER:
   nil
ARGS:

PROTECTED CALL STACK:
	Meta_MethodError:new	0x3dc0600
		arg this = DoesNotUnderstandError
		arg what = nil
		arg receiver = nil
	Meta_DoesNotUnderstandError:new	0x3dc25c0
		arg this = DoesNotUnderstandError
		arg receiver = nil
		arg selector = defaultGroup
		arg args = [  ]
	Object:doesNotUnderstand	0x1fe38c0
		arg this = nil
		arg selector = defaultGroup
		arg args = nil
	a FunctionDef	0x5146800
		sourceCode = "{ ~server.defaultGroup.nodeID }"
	a FunctionDef	0x5164100
		sourceCode = "#{ |server|
						var bndl, id, ids, addAction, f;
						addAction = Node.actionNumberFor(~addAction);
						~freq = ~detunedFreq.value;
						f = ~freq;
						~amp = ~amp.value;

						bndl = ( [9 /* \\s_new */, ~instrument, ids, addAction, ~group.asControlInput]
							++ ~msgFunc.valueEnvir).flop;
						bndl.do { | b |
							id = server.nextNodeID;
							ids = ids.add(id);
							b[2] = id;
						};
						~id = ids;

						if ((addAction == 0) || (addAction == 3)) {
							bndl = bndl.r...etc..."
		arg server = localhost
		var bndl = nil
		var id = nil
		var ids = nil
		var addAction = 0
		var f = 261.6255653006
	a FunctionDef	0x5152c40
		sourceCode = "#{
					var tempo, server, eventTypes, parentType;

					parentType = ~parentTypes[~type];
					parentType !? { currentEnvironment.parent = parentType };

					~finish.value(currentEnvironment);

					server = ~server ? Server.default;

					tempo = ~tempo;
					tempo !? { thisThread.clock.tempo = tempo };


					if(currentEnvironment.isRest.not) {
						eventTypes = ~eventTypes;
						(eventTypes[~type] ?? { eventTypes[\\note] }).value(server)
					};

					~callback.value(currentEnvironmen...etc..."
		var tempo = nil
		var server = localhost
		var eventTypes = ( 'fadeBus': a Function, 'freeAllocWrite': a Function, 'tree': a Function, 'on': a Function, 
  'load': a Function, 'freeBuffer': a Function, 'group': a Function, 'freeAllocRead': a Function, 'allocWrite': a Function, 
  'cue': a Function, 'grain': a Function, 'Synth': a Function, 'freeAllocWriteID': a Function, 'rest': a Function, 
  'alloc': a Function, 'sine2': a Function, 'sine1': a Function, 'midi': a Function, 'set': a Function, 
  'setProperties': a Function, 'parGroup': a Function, 'allocRead': a Fu...etc...
		var parentType = nil
	a FunctionDef	0x5105700
		sourceCode = "<an open Function>"
	Function:prTry	0x37bbe40
		arg this = a Function
		var result = nil
		var thread = a Routine
		var next = nil
		var wasInProtectedFunc = false
	
CALL STACK:
	DoesNotUnderstandError:reportError
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError
		arg this = <instance of Routine>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw
		arg this = <instance of DoesNotUnderstandError>
	Function:protect
		arg this = <instance of Function>
		arg handler = <instance of Function>
		var result = <instance of DoesNotUnderstandError>
	Environment:use
		arg this = <instance of Event>
		arg function = <instance of Function>
		var result = nil
		var saveEnvir = <instance of Environment>
	Event:play
		arg this = <instance of Event>
	Event:playAndDelta
		arg this = <instance of Event>
		arg cleanup = <instance of EventStreamCleanup>
		arg mute = false
	EventStreamPlayer:prNext
		arg this = <instance of EventStreamPlayer>
		arg inTime = 8.846657349
		var nextTime = nil
		var outEvent = <instance of Event>
	< FunctionDef in Method EventStreamPlayer:init >
		arg inTime = 8.846657349
	Routine:prStart
		arg this = <instance of Routine>
		arg inval = 8.846657349
^^ The preceding error dump is for ERROR: Message 'defaultGroup' not understood.
RECEIVER: nil
```

**Comment by jamshark70** [comment]
I'm surprised this doesn't happen with normal event types.

Because the default `~group` is now a function `{ ~server.defaultGroupID }`, we're relying on `asOSCArg...` methods to call `asControlInput` on the function and get the group ID.

Also, event types generally defer sending OSC through some scheduled functions (to handle timingOffset and lag).

What *should* happen is:

1. Event type play function generates a bundle (which now includes the group function).
2. Event type function processes the bundle (some are using `.collect(_.asOSCArgArray)`, others `asOSCArgBundle` -- I'm not sure if this makes a difference). This should resolve the group function into an integer ID.
3. Event type function schedules the completed bundle, which by now should be free of all non-OSC-able entities, for sending.

It looks to me like the actual sequence in this issue skips step 2, and processes the bundle in a scheduled function. Because it's in a scheduled function, you're no longer in the event's environment, so `~server` is invalid (nil).

At first glance, I don't see why it (I guess) is working for event type `\note` but not for the mono event types.

Possibly a better solution would be to avoid the sleight-of-hand with the function -- to be honest, I don't trust it, it's "clever" but I grow to dislike "clever" solutions more and more, as they open up unforeseen consequences -- and:

- Remove `group` from the default event.
- Change `Event:play` to fill in the default group if the user hasn't overridden.

**Comment by jamshark70** [comment]
But I really don't get it because `monoNote` says:

```
bndl = ( [9 /* \s_new */, ~instrument, ids, addAction, ~group.asControlInput]
```

So I'm baffled.

Also, a curious discrepancy: the referenced PR defines `{ ~server.defaultGroupID }`, but the stack trace shows that it's trying to evaluate a function `{ ~server.defaultGroup.nodeID }` -- so, either:

- `group` is being changed somewhere (I reviewed my code for PmonoStreams, it isn't happening there)

- or the issue may be happening in a version that hasn't fully integrated the PR.

**Comment by patrickdupuis** [comment]
Thanks for looking at this @jamshark70. If you're baffled, then we really are screwed with this one.

It's not clear whether or not you reproduced the bug on your system. I'm pretty certain that the commit that I reference is the point at which this breakage occurred. However, I'm not a git wizard. It would be good if someone could confirm this.

**Comment by jamshark70** [comment]
I haven't tried to reproduce it (yet) because, after replacing my computer after the boat accident in Thailand, I took the quickest way to get back to my previous working environment -- copying the repository from my backup hard drive. Unfortunately, it seems git now thinks that literally every file has local changes, so now...

```
$ git fetch upstream
... OK
$ git checkout 3.9

error: Your local changes to the following files would be overwritten by checkout:
	detail/vec_math.hpp
Please, commit your changes or stash them before you can switch branches.
Aborting
Unable to checkout 'd2c87d6a9776e857b54c6bb4533551034e4a1bc1' in submodule path 'external_libraries/nova-simd'

$ git diff external_libraries/nova-simd/detail/vec_math.hpp
$
```

... and git diff reports *there aren't any local changes*, but I still can't check out the branch.

So, I've got to waste a ton of time in the next couple of weeks resolving a visa situation, I have to finish a new piece for a recital in three weeks, and git is doing THIS???

I guess I'll have to check out a fresh clone, but... I'm getting quite tired of git. Every time I sneeze, it says I'm doing something wrong.

**Comment by telephon** [comment]
I can't reproduce this yet. I've tried

```
Pmono(\default, \dur, 0.2, \freq, Pwhite(1,8) * 100 ).play
```

that works fine.

**Comment by patrickdupuis** [comment]
@telephon you are on the latest 3.9? I think that `develop` will not have the changes from this commit.

@jamshark70 take your time getting your system setup again. No worries :)

**Comment by jamshark70** [comment]
OK, now I see what it is.

```
Event.default[\server]
-> nil
```

But:

```
				group: { ~server.defaultGroup.nodeID },
```

We are assuming that the event will contain something under `~server` but this is not actually happening.

In the default event's `play` function, there is:

```
					server = ~server ? Server.default;
```

... which ensures that the server will be passed into the event type's function, but if the event doesn't specify a server, it doesn't fix the event to include one.

The `\note` event type does `~server = server;` so... nobody noticed that this is the wrong place to fix up the event with the server variable. That should be done in `play: { ... }`.

BTW, when I first started testing, I saw something *very weird* in that an event containing a SynthDesc object was identified as a rest???? But, after stopping and rebooting the interpreter, the problem went away. That's troubling... it means it's possible to get the interpreter into a condition where the isRest primitive is mis-identifying an object's class. But I didn't see that problem again...
